### PR TITLE
feat: use Workers for manifest fetching & optimizations

### PR DESCRIPTION
**Pull request by kyrie25**
- Use Web Workers to fetch manifest contents, this way the console will no longer be littered with 404 spams.
image

- Reworked & fixed cache system (so that it actually works now), no longer making duplicate requests and decrease load time.

- Topic results will also be cached and will no longer cause rate limits when spamming switching tabs.

- A notification will now appear if all JSDelivr endpoints can't be connected to (except when they are offline, in which the extension will reload when online)
image

### CODE REVIEW

File: src/logic/FetchRemotes.ts [hunk 1]
```diff
@@ -83,7 +98,7 @@ export async function fetchExtensionManifest(contents_url: string, branch: strin
     if (!regex_result || !regex_result.groups) return null;
     const { user, repo } = regex_result.groups;
 
-    manifests = await getRepoManifest(user, repo, branch);
+    manifests = await getRepoManifest(user, repo, regex_result[0]);
```
Comment (theRealPadster) [inline src/logic/FetchRemotes.ts:101]: Why is this using regex result instead of branch?

File: src/logic/FetchRemotes.ts [hunk 2]
```diff
@@ -130,8 +145,8 @@ export async function fetchExtensionManifest(contents_url: string, branch: strin
 
     return parsedManifests;
   }
-  catch (err) {
-    // console.warn(contents_url, err);
+  catch (err: any) {
```
Comment (theRealPadster) [inline src/logic/FetchRemotes.ts:148]: Types

File: src/types/marketplace-types.d.ts [hunk 3]
```diff
@@ -162,3 +162,25 @@ export type Config = {
...
+    self: string;
+    git: string;
+    html: string;
+  };
+};
+
+export type GithubMessage = {
+  message: string;
+  documentation_url: string;
+};
```
Comment (theRealPadster) [inline src/types/marketplace-types.d.ts:186]: Missing EOF newline

**Review by theRealPadster — CHANGES_REQUESTED** [review]

File: src/logic/FetchRemotes.ts (hunk 2 above)
Comment (kyrie25) [reply src/logic/FetchRemotes.ts:148]: You can only assign either unknown or any to caught errors and type check is done in the catch block anyway.

File: src/logic/FetchRemotes.ts [hunk 4]
```diff
@@ -51,17 +83,20 @@ export async function getTaggedRepos(tag: RepoTopic, page = 1, BLACKLIST:string[
 * @returns The manifest object
 */
 async function getRepoManifest(user: string, repo: string, branch: string) {
-  const sessionStorageItem = window.sessionStorage.getItem(`${user}-${repo}`);
-  const failedSessionStorageItems = window.sessionStorage.getItem("noManifests");
+  const key = `${user}-${repo}`;
```
Comment (theRealPadster) [inline src/logic/FetchRemotes.ts:86]: Do we not need the branch anymore? What if the user has one repo and uses different branches for different projects?

Comment (kyrie25) [reply src/logic/FetchRemotes.ts:86]: That's still the same key we used before, just initialized once.
Also one repo only has one manifest. This caches everything in the manifest file, not one manifest object itself.

File: src/components/Card/Card.tsx [hunk 5]
```diff
@@ -379,7 +378,7 @@ class Card extends React.Component<CardProps, {
    */
   async fetchAndInjectUserCSS(theme) {
     try {
-      const tld = await getAvailableTLD();
+      const tld = window.sessionStorage.getItem("tld") || "";
```
Comment (rxri) [inline src/components/Card/Card.tsx:381]:
```suggestion
      const tld = window.sessionStorage.getItem("marketplace-request-tld") || undefined;
```

File: src/logic/Utils.ts [hunk 6]
```diff
@@ -417,9 +417,11 @@ export const initAlbumArtBasedColor = (scheme: ColourScheme) => {
   });
 };
 
-export const parseCSS = async (themeData: CardItem, tld = "net") => {
+export const parseCSS = async (themeData: CardItem, tld?: string) => {
```
Comment (rxri) [inline src/logic/Utils.ts:420]:
```suggestion
export const parseCSS = async (themeData: CardItem, tld?: string | undefined) => {
```

**Review by rxri — CHANGES_REQUESTED** [review]

File: src/logic/Utils.ts (hunk 6 above)
Comment (kyrie25) [reply src/logic/Utils.ts:420]: Optional param is explicitly undefined

**Review by theRealPadster — APPROVED** [review]

**Review by rxri — APPROVED** [review]

**Pull request merged**